helpers/misc.go

```diff
@@ -18,26 +19,44 @@ func BoolPtr(b bool) *bool {
 	return &b
 }
 
-// Retry will retry a function until it succeeds or the timeout is reached. timeout == 2^attempt * delay.
-func Retry(fn func() error, retries int, delay time.Duration, logger func(format string, args ...any)) error {
+// RetryWithContext will retry a function until it succeeds, the timeout is reached, or context is done. timeout == 2^attempt * delay.
+func RetryWithContext(ctx context.Context, fn func() error, retries int, delay time.Duration, logger func(format string, args ...any)) error {
 	var err error
 	for r := 0; r < retries; r++ {
-		err = fn()
-		if err == nil {
-			break
-		}
-
-		pow := math.Pow(2, float64(r))
-		backoff := delay * time.Duration(pow)
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+
+			err = fn()
+			if err == nil {
+				break
+			}
+			pow := math.Pow(2, float64(r))
+			backoff := delay * time.Duration(pow)
 
-		logger("Retrying (%d/%d) in %s: %s", r+1, retries, backoff, err.Error())
+			logger("Retrying (%d/%d) in %s: %s", r+1, retries, backoff, err.Error())
 
-		time.Sleep(backoff)
+			timer := time.NewTimer(backoff)
+			select {
+			case <-timer.C:
+			case <-ctx.Done():
+				if !timer.Stop() {
+					<-timer.C
+				}
+				return ctx.Err()
+			}
+		}
 	}
 
 	return err
 }
```

Comment:
I think we can do a little bit of simplifying of this logic. Here are my suggestions:

- only have 1 `select` statement in this function (still within the `for` loop)
- iterate from `1 -> retries`

example:
```go
for attempt := 1; attempt <= retries; attempt++ {
    // fn()
    
    // backoff calc
    // log message
    
    // new timer
    // select statement   
}

return fmt.Errorf("all attempts failed after %d retries", retries)
	
```

Reply:
Kinda confused at the suggestion. Isn't `for attempt := 1; attempt <= retries;` and `for attempt := 0; attempt < retries;` exactly the same? The reason for the first select block is so if ctx is finished is called before the function is called the function won't go through. The reason for the second is so we can skip the timeout in the event ctx is done

Reply:
> Kinda confused at the suggestion. Isn't for attempt := 1; attempt <= retries; and for attempt := 0; attempt < retries

Behavior wise, yes. This is not a behavior issue or a bug, but rather a style/readability/simplicity suggestion (which is mostly subjective). Seems odd to me to start from `0` for this use case since we aren't indexing an array, and we are having to account for this with `r+1` in the log message. If you feel strongly about wanting to keep the starting point for the iteration at `0`, then I'm okay with that.

> The reason for the first select block is so if ctx is finished is called before the function is called the function won't go through. The reason for the second is so we can skip the timeout in the event ctx is done

Makes sense, thanks for the explanation. Disregard the invalid suggestion to use only one `select` statement